hudi-aws/src/main/java/org/apache/hudi/aws/transaction/lock/S3StorageLockClient.java

```diff
@@ -0,0 +1,248 @@
+ */
+
+package org.apache.hudi.aws.transaction.lock;
+
+import org.apache.hudi.aws.credentials.HoodieAWSCredentialsProviderFactory;
+import org.apache.hudi.client.transaction.lock.models.LockGetResult;
+import org.apache.hudi.client.transaction.lock.models.LockUpdateResult;
+import org.apache.hudi.client.transaction.lock.StorageLock;
+import org.apache.hudi.client.transaction.lock.models.StorageLockData;
+import org.apache.hudi.client.transaction.lock.models.StorageLockFile;
+import org.apache.hudi.common.util.VisibleForTesting;
+import org.apache.hudi.common.util.collection.Pair;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.awscore.exception.AwsServiceException;
+import software.amazon.awssdk.core.ResponseInputStream;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.regions.providers.DefaultAwsRegionProviderChain;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Properties;
+import java.util.function.Supplier;
+
+/**
+ * S3-based distributed lock service using ETag checks (AWS SDK v2).
+ * See RFC: <a href="https://github.com/apache/hudi/blob/master/rfc/rfc-91/rfc-91.md">RFC-91</a>
+ */
+@ThreadSafe
+public class S3StorageLockClient implements StorageLock {
+
+  private static final Logger LOG = LoggerFactory.getLogger(S3StorageLockClient.class);
+  private static final long WAIT_TIME_FOR_RETRY_MS = 1000L;
+  private static final int PRECONDITION_FAILURE_ERROR_CODE = 412;
+  private static final int NOT_FOUND_ERROR_CODE = 404;
+  private static final int CONDITIONAL_REQUEST_CONFLICT_ERROR_CODE = 409;
+  private static final int RATE_LIMIT_ERROR_CODE = 429;
+  private static final int INTERNAL_SERVER_ERROR_CODE_MIN = 500;
+
+  private final Logger logger;
+  private final S3Client s3Client;
+  private final String bucketName;
+  private final String lockFilePath;
+  private final String ownerId;
+
+  /** Constructor that is used by reflection to instantiate an S3-based storage locking client.
+   * @param ownerId The owner id.
+   * @param bucketName The name of the bucket.
+   * @param lockFilePath The path within the bucket where to write lock files.
+   * @param props The properties for the lock config, can be used to customize client.
+   */
+  public S3StorageLockClient(String ownerId, String bucketName, String lockFilePath, Properties props) {
+    this(
```

Comment:
style

Reply:
My linter must be broken, I'm not sure what the indentation should be, I will get this fixed with the next set of feedback that lands.